src/executor/evaluate/mod.rs

```diff
@@ -170,6 +172,74 @@ pub async fn evaluate<'a, T: 'static + Debug>(
         Expr::Wildcard | Expr::QualifiedWildcard(_) => {
             Err(EvaluateError::UnreachableWildcardExpr.into())
         }
+        Expr::Case {
+            operand,
+            when_then,
+            else_result,
+        } => {
+            let when_then: Vec<(Evaluated, Evaluated)> = stream::iter(when_then.iter())
+                .map(Ok::<_, Error>)
+                .and_then(|(when, then)| async move { Ok((eval(when).await?, eval(then).await?)) })
+                .try_collect()
+                .await?;
+
+            let else_result = match else_result {
+                Some(result) => Some(eval(result).await?),
+                None => None,
+            };
+
+            let parse_pair =
+                |result_pair: (Result<Value>, Result<Value>)| -> Result<(Value, Value)> {
+                    match result_pair {
+                        (Err(e), _) | (_, Err(e)) => Err(e),
+                        _ => Ok((result_pair.0.unwrap(), result_pair.1.unwrap())),
+                    }
+                };
+
+            let when_then_val = when_then
+                .clone()
+                .into_iter()
+                .map(|(when, then)| parse_pair((when.try_into(), then.try_into())))
+                .collect::<Result<Vec<(Value, Value)>>>()?;
+
+            if let Some(er) = else_result {
+                let else_result = er.try_into()?;
+                let then_type = when_then_val
+                    .iter()
+                    .map(|(_, t)| (discriminant(t)))
+                    .unique()
+                    .collect_vec();
+
+                if let Some(then_type_first) = then_type.first() {
+                    if then_type_first != &discriminant(&else_result) || then_type.len() != 1 {
+                        return Err(EvaluateError::UnequalResultTypes("CASE".to_owned()).into());
+                    }
+                }
+            }
```

Comment:
This part checks `when` result type equality only when `else_result` exists, but it should check also when the `else_result` is `None`. Just a self-review for my mistake!